Add EIP: Order Book DEX with Two Phase Withdrawal (#6787)
ethereum/EIPs: EIPS/eip-6787.md

## Specification

The Order Book DEX Standard defines the following functions:

### `deposit`

`function deposit(address token, uint256 amount) external;`

The **deposit** function allows a user to deposit a specified amount of a particular token to the exchange. The *token* parameter specifies the address of the token contract, and the *amount* parameter specifies the amount of the token to be deposited.

### `withdraw`

`function withdraw(address token, uint256 amount) external;`

The **withdraw** function allows a user to withdraw a specified amount of a particular token from the exchange. The *token* parameter specifies the address of the token contract, and the *amount* parameter specifies the amount of the token to be withdrawn.

### `settle`

> The settle interface has different implementations for different order book exchanges, and here is an example. When inheriting the business, you can also perform settle operations for batch users or multiple currencies.

Author: Yes, `settle` can have different signatures. This is the example of two people and two currency. But this also the recommend way for implementer.

> The settle interface has different implementations for different order book exchanges, and here is an example. When implements the business, you can also perform settle operations for batch users or multiple currencies.

Reviewer: > When implements the business

I think this is a mis-translation? Perhaps:

```suggestion
The settle interface has different implementations for different order book exchanges, and here is an example. When implemented in your application, you can also perform settle operations for batch users or multiple currencies.
```

> ### `settle`
> 
> The settle interface has different implementations for different order book exchanges, and here is an example. When implements the business, you can also perform settle operations for batch users or multiple currencies.
> However, two points need to be noted:
> 
> 1. The settle interface can only be called by the exchange;
> 2. The settle implementation needs to verify user balances and balance all involved users and currency balances, similar to the balance checking function of UTXO input and output.
> 
> Here is the example.
> 
> `function settle(address userA, address tokenA, uint256 amountA, address userB, address tokenB, uint256 amountB) external;`
> 
> The example **settle** function is used to settle transactions between two parties, involving the exchange of assets represented by tokenA and tokenB. The function takes the following parameters:
> 
> 1. *userA*: the address of the first user involved in the transaction
> 2. *tokenA*: the address of the token being traded by user A
> 3. *amountA*: the amount of token A being traded by user A
> 4. *userB*: the address of the second user involved in the transaction
> 5. *tokenB*: the address of the token being traded by user B
> 6. *amountB*: the amount of token B being traded by user B

Reviewer: Since `settle` is supposed to be called by the exchange, and not by users or other contracts, it shouldn't be part of the standard. This is very similar to how [ERC-20](https://eips.ethereum.org/EIPS/eip-20) and [ERC-721](https://eips.ethereum.org/EIPS/eip-721) don't specify a `mint` function. You can (and should) still explain how settling works, and if you add a reference implementation, `settle` should probably be there.

```suggestion
```

Author: I see. Thank you for your guidance. I will make some adjustments.

By introducing the two-phase forced withdrawal function, we can protect users' financial security while ensuring the security of the exchange's assets. In the first phase, the system will conduct a preliminary review of the user's withdrawal request to confirm the user's account status. In the second phase, after the forced withdrawal inspection period, users can directly submit the forced withdrawal request to complete the forced withdrawal process. In this way, we can not only prevent users from exploiting the forced withdrawal function to evade losses but also ensure the asset security for both the exchange and the users.

In conclusion, by adopting the two-phase commit protocol and the two-phase forced withdrawal function, we can effectively guard against malicious behaviors and ensure data consistency and security in distributed systems while ensuring user experience and financial security.

> The flow charts for two-phase withdrawal are shown below：

Reviewer: ```suggestion
The flow charts for two-phase withdrawal are shown below:
```

However, such design also faces potential risks. When users engage in perpetual contract transactions, they may incur losses. In this situation, malicious users might exploit the forced withdrawal function to evade losses. To prevent this kind of attack, we propose a two-phase forced withdrawal mechanism.

By introducing the two-phase forced withdrawal function, we can protect users' financial security while ensuring the security of the exchange's assets. In the first phase, the system will conduct a preliminary review of the user's withdrawal request to confirm the user's account status. In the second phase, after the forced withdrawal inspection period, users can directly submit the forced withdrawal request to complete the forced withdrawal process. In this way, we can not only prevent users from exploiting the forced withdrawal function to evade losses but also ensure the asset security for both the exchange and the users.

> In conclusion, by adopting the two-phase commit protocol and the two-phase forced withdrawal function, we can effectively guard against malicious behaviors and ensure data consistency and security in distributed systems while ensuring user experience and financial security.

Reviewer: So this looks like it's justifying why we need two-phase withdrawals, which is the point of the whole EIP. This kind of information belongs in the motivation section.

The rationale should describe individual technical choices made within the EIP itself.

Author: Yes, an account may initiate multiple forced withdrawal requests in parallel, so requestid is required.

Author: Yes, an account may initiate multiple forced withdrawal requests in parallel, so requestid is required.

Reviewer: Perfect. If you move the current rationale into the motivation section and add that note about `requestId` in the rationale, that should be it before merging as draft.

Author: Thank you for your reply, I have fix it.

---
eip: 6787
title: Order Book DEX with Two Phase Withdrawal

> description: A order book-based DEX Interfaces that ensures the asset security of both users and the exchange

Reviewer: ```suggestion
description: An order book-based DEX interface that ensures the asset security of both users and the exchange
```

### withdraw

`function withdraw(address token, uint256 amount) external;`

The **withdraw** function allows a user to withdraw a specified amount of a particular token from the exchange. The *token* parameter specifies the address of the token contract, and the *amount* parameter specifies the amount of the token to be withdrawn.

### settle

`function settle(address userA, address tokenA, uint256 amountA, address userB, address tokenB, uint256 amountB) external;`

The **settle** function is used to settle transactions between two parties, involving the exchange of assets represented by tokenA and tokenB. The function takes the following parameters:

1. *userA*: the address of the first user involved in the transaction
2. *tokenA*: the address of the token being traded by user A
3. *amountA*: the amount of token A being traded by user A
4. *userB*: the address of the second user involved in the transaction
5. *tokenB*: the address of the token being traded by user B

> 6. *amountB*: the amount of token B being traded by user B

Author: Thank for your comment! The settle interface has different implementations for different order book exchanges, and here is an example. I just provided a more detailed explanation in the proposal.

1. *token*: the address of the token to be withdrawn
2. *amount*: the amount of the token to be withdrawn

The function returns a unique *requestID* that can be used to submit the forced withdrawal using the commitForceWithdraw function.

### commitForceWithdraw

`function commitForceWithdraw(uint256 requestID) external;`

The **commitForceWithdraw** function is used to execute a forced withdrawal operation after the conditions are met. The function takes a *requestID* parameter, which specifies the ID of the forced withdrawal request to be executed. The request must have been previously initiated using the prepareForceWithdraw function.

The flow charts for **prepareForceWithdraw** and **commitForceWithdraw** are shown below：

![](../assets/eip-6787/image1.png)

## Rationale

> 

Reviewer: You'll need to include some explanation for any technical choices made within this proposal.

## Backwards Compatibility

No backward compatibility issues found.

> ## Test Cases
> 
> 
> ## Reference Implementation
> 
> 

Reviewer: ```suggestion
```

These sections can be omitted if empty.
